pkg/util/workqueue/delaying_queue.go

```diff
@@ -0,0 +1,190 @@
+	ret := &delayingType{
+		Type:        New(),
+		clock:       clock,
+		waitingCond: sync.NewCond(&sync.Mutex{}),
+	}
+
+	go ret.waitingLoop()
+	ret.notifyAt(clock.Now().Add(maxWait))
+
+	return ret
+}
+
+// delayingType wraps a Type and provides delayed re-enquing
+type delayingType struct {
+	*Type
+
+	// clock tracks time for delayed firing
+	clock util.Clock
+
+	// waitingForAdd is an ordered slice of items to be added to the contained work queue
+	waitingForAdd []waitFor
+	// waitingLock synchronizes access to waitingForAdd
+	waitingLock sync.Mutex
+	// waitingCond is used to notify the adding go func that it needs to check for items to add
+	waitingCond *sync.Cond
+
+	// nextCheckTime is used to decide whether to add a notification timer.  If the requested time
+	// is beyond the time we're already waiting for, we don't add a new timer thread
+	nextCheckTime *time.Time
+	// nextCheckLock serializes access to the notification time
+	nextCheckLock sync.Mutex
+}
+
+// waitFor holds the data to add and the time it should be added
+type waitFor struct {
+	data    t
+	readyAt time.Time
+}
+
+// ShutDown gives a way to shut off this queue
+func (q *delayingType) ShutDown() {
+	q.Type.ShutDown()
+	q.waitingCond.Broadcast()
+}
+
+func (q *delayingType) AddAfter(item interface{}, duration time.Duration) {
+	q.waitingLock.Lock()
+	defer q.waitingLock.Unlock()
+	waitEntry := waitFor{data: item, readyAt: q.clock.Now().Add(duration)}
+
+	insertionIndex := sort.Search(len(q.waitingForAdd), func(i int) bool {
+		return waitEntry.readyAt.Before(q.waitingForAdd[i].readyAt)
+	})
+
+	tail := q.waitingForAdd[insertionIndex:]
+	q.waitingForAdd = append(make([]waitFor, 0, len(q.waitingForAdd)+1), q.waitingForAdd[:insertionIndex]...)
+	q.waitingForAdd = append(q.waitingForAdd, waitEntry)
+	q.waitingForAdd = append(q.waitingForAdd, tail...)
+
+	q.notifyAt(waitEntry.readyAt)
+}
+
+// maxWait keeps a max bound on the wait time.  It's just insurance against weird things happening.
+// Checking the queue every 10 seconds isn't expensive and we know that we'll never end up with an
+// expired item sitting for more than 10 seconds.
+const maxWait = 10 * time.Second
+
+// waitingLoop runs until the workqueue is shutdown and keeps a check on the list of items to be added.
+func (q *delayingType) waitingLoop() {
+	defer utilruntime.HandleCrash()
+
+	for {
+		if q.shuttingDown {
+			return
+		}
+
+		func() {
+			q.waitingCond.L.Lock()
+			defer q.waitingCond.L.Unlock()
+			q.waitingCond.Wait()
+
+			if q.shuttingDown {
+				return
+			}
+
+			q.waitingLock.Lock()
+			defer q.waitingLock.Unlock()
+
+			// allow the next notifyAt to take effect.  At this instant, we have no notification time
+			q.clearNextCheckTime()
+
+			now := q.clock.Now()
+
+			// always check at least every 10 seconds
+			nextReadyCheck := now.Add(maxWait)
+			nextItemIndex := 0
+
+			for i, queuedItem := range q.waitingForAdd {
+				nextItemIndex = i
```

Comment:
as written, this will always leave at least one item in `waitingForAdd`, and `nextItemIndex` doesn't distinguish between the case where we `break` (and should retain the item) and the case where we `q.Type.Add(...)` (and should remove the item)